### ISSUE TITLE
[Ray Cluster Launcher] tpu pod support

### ISSUE THREAD

**Issue by JiahaoYao** [ray-project/ray#25223]
### Description

**Current case**: the autoscaler supports TPU but only `v2-8` not tpu pod, i.e. `v2-32`. 

See details here:
https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/_private/gcp/config.py#L72-L76

**Request**: support the tpu pod autoscaler 

**difficulty**: is TPU pods are nodes with multiple sub-nodes (multiple ip addresses). For example, a tpu pod `v2-32` is going to have 4 ip addresses, while the current autoscaler. assumes each new vm has only one ip address.

<img width="1182" alt="image" src="https://user-images.githubusercontent.com/20907377/170521393-35b60137-9cec-43c6-a893-ee2de2050a53.png">

@Yard1 @DmitriGekhtman

**Comment by DmitriGekhtman** [comment]
@JiahaoYao
I'm curious if the multiple ips cause issues for Ray components other than the autoscaler.
Would it be possible to manually start (`ray start`) a Ray cluster with a CPU head and a v-32 TPU worker and correctly use the TPUs?

@JiahaoYao @Yard1 could you assign a priority level (P1 or P2) to this?

BTW, thanks for opening the issue -- it's good to have actual GitHub issues for our todos :)

**Comment by Yard1** [comment]
It is possible to manually ssh into each separate IP and start ray there. It works fine. The only problem is with autoscaler not being able to handle this setup.

We are seeing more and more user asks for this. I'll defer to @JiahaoYao on whether this is a blocker or not :)

**Comment by DmitriGekhtman** [comment]
> each separate IP

I see, so we have a Ray node per TPU subnode.
Can GCPNodeProvider be augmented to handle this situation, so that the autoscaler considers each subnode a separate Ray node?

**Comment by JiahaoYao** [comment]
> I'm curious if the multiple ips cause issues for Ray components other than the autoscaler.
> Would it be possible to manually start (`ray start`) a Ray cluster with a CPU head and a v-32 TPU worker and correctly use the TPUs?

@DmitriGekhtman thanks for your prompt response. @Yard1 thanks for adding more info here, appreciate it!


> Would it be possible to manually start (ray start) a Ray cluster with a CPU head and a v-32 TPU worker and correctly use the TPUs?

Yes, as @Yard1 said, i was using the following command to manually enable the ray cluster on the tpu pod 

```python 
gcloud alpha compute tpus tpu-vm ssh jax-trainer-mnist-tpu-pod --zone=us-central1-a --command "ray stop && ray start --head --port=6379 --resources='{\"TPU\":1}'" --worker=0

gcloud alpha compute tpus tpu-vm ssh jax-trainer-mnist-tpu-pod --zone=us-central1-a --command "ray stop && ray start --address='10.128.0.38:6379' --resources='{\"TPU\":1}'" --worker=1

gcloud alpha compute tpus tpu-vm ssh jax-trainer-mnist-tpu-pod --zone=us-central1-a --command "ray stop && ray start --address='10.128.0.38:6379' --resources='{\"TPU\":1}'" --worker=2

gcloud alpha compute tpus tpu-vm ssh jax-trainer-mnist-tpu-pod --zone=us-central1-a --command "ray stop && ray start --address='10.128.0.38:6379' --resources='{\"TPU\":1}'" --worker=3
```

As for the priority level, we have been seeing more and more ppl asking for the support on TPU for the ray train and others. So, it would be nice if this can be P1 issue.

**Comment by JiahaoYao** [comment]
> Can GCPNodeProvider be augmented to handle this situation, so that the autoscaler considers each subnode a separate Ray node?

Yes, this is a nice point. And I am also thinking along this line!

**Comment by DmitriGekhtman** [comment]
> > Can GCPNodeProvider be augmented to handle this situation, so that the autoscaler considers each subnode a separate Ray node?
> 
> Yes, this is a nice point. And I am also thinking along this line!

Following up on this, we need the autoscaler to account one "node" in its internal processing per actual Ray node. 

The necessary modifications to GCPNodeProvider are probably not completely trivial -- the infra team has also not allocated time in June planning for this.
@Yard1 @JiahaoYao would either of you have bandwidth to try to implement this support?

Alternatively, infra team could consider this as a potential item for July.
cc @AmeerHajAli @clarng @yifeif for planning context

**Comment by DmitriGekhtman** [comment]
What's the fundamental advantage of running multiple Ray nodes on one TPU pod?
Are there shared hardware resources the collocated Ray nodes can take advantage of? 
Is it possible to run a Ray node at just one of the subnodes and somehow take advantage of all of the TPU resources at the rest of the nodes?

**Comment by JiahaoYao** [comment]
These are really good questions!

> Is it possible to run a Ray node at just one of the subnodes and somehow take advantage of all of the TPU resources at the rest of the subnodes?

I tried this before, but didn't work. 

For example, using the gcp cli, the [command to launch the job](https://cloud.google.com/tpu/docs/jax-pods) is 
```python
gcloud compute tpus tpu-vm ssh tpu-name \
  --zone europe-west4-a --worker=all --command "python3 example.py"
```

Note `--worker=all` here, this makes the command run on each subnodes, or called [Single Program, Multiple Data (SPMD)](https://jax.readthedocs.io/en/latest/multi_process.html). 


<img width="659" alt="image" src="https://user-images.githubusercontent.com/20907377/170573641-ab8e8d95-b9dd-4a2b-819a-fe1e4812a880.png">

And, in my understanding, the setup of the ray cluster is to actually only launch the job on one of the subnodes (head-node) and take all the TPU resources. I mean after setting up the `ray up` in the autoscalar, the head node can use all the computation.

**Comment by JiahaoYao** [comment]
<img width="909" alt="image" src="https://user-images.githubusercontent.com/20907377/170584122-a3db3dd0-5954-4b73-9d32-b4b262783283.png">

see [here](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#:~:text=Tensor%20Processing%20Units%20(TPUs)%20are,TensorFlow%2C%20Pytorch%2C%20and%20JAX.) for the detail of image. 

Each `v2-32` contains 4 `v2-8`. 
Yes, I was also thinking about this question and submitted an issue at Jax https://github.com/google/jax/discussions/10802. Long story short, for the latter, we can directly launch the ray clusters using autoscaler on multiple `v2-8`. Then the question is the communication between the nodes. this is still question marks. In theory, there should be some codes to setup the connections for XLA, though I hear the communication speed of this is slower than the tpu pod, where the connections are physical from the above image.

**Comment by DmitriGekhtman** [comment]
Nice, thanks for the context!

It sounds like we still need to resolve questions around TPU board communication -- these are questions at the level of the Ray runtime and Ray libraries. Once we figure out how to do things on a fixed TPU Ray cluster, we can start to think about how the autoscaler should provision resources.

**Comment by JiahaoYao** [comment]
Cool, thanks for your in-depth thoughts on this, really appreciate it !

**Comment by OhadRubin** [comment]
This would be a nice to have!
Thank you!

**Comment by OrenLeung** [comment]
@jiaodong any updates on tpu pod autoscaler?

**Comment by OrenLeung** [comment]
ray cluster launcher for tpu vm pods would help my life a lot!

**Comment by DmitriGekhtman** [comment]
There's a hacky implementation [in this branch](https://github.com/DmitriGekhtman/ray/tree/hackathon/tpu_pod) (implemented during an Anyscale hackathon)

Some strange things had to be done to allow the autoscaler to understand that 1 instance can correspond to multiple VMs
https://github.com/DmitriGekhtman/ray/blob/hackathon/tpu_pod/python/ray/autoscaler/_private/gcp/node_provider.py#L118-L130
The implementation is not at all production ready and some functionality is missing
(for example, chaos will ensue if the autoscaler tries to terminate a VM in a TPU pod)

**Comment by DmitriGekhtman** [comment]
I think the only sane way to do this would be to adapt [KubeRay](https://github.com/ray-project/kuberay) to work with [GKE's TPU support](https://cloud.google.com/tpu/docs/kubernetes-engine-setup). This would only work well with statically-sized clusters (Ray autoscaler disabled).

**Comment by OrenLeung** [comment]
@DmitriGekhtman thanks for the followup!

i assume that there would be a non-trivial code change to get it so that 1 tpu pod slice has multiple VMs associated it.

TPU VMs are not currently supported on GKE. The link you provided is for the TPU node architecture. [explainer on tpu node vs tpu VM](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu-arch)

according to google, TPU VM on GKE will hit alpha around early 2023. https://github.com/google/jax/issues/12917#issuecomment-1287404101

in regards to yr comment about autoscaler, auto scaling will need to consider even more things as tpu vms due to its topology, they need to gang schedule jobs on the same pod slice.

i think just cluster launcher without autoscaler would be benefit to the community. i know a few companies and jax/tpu users whom will use this.

**Comment by DmitriGekhtman** [comment]
Waiting for alpha support from GKE may be the best route.

**Comment by kevin85421** [comment]
In my understanding, this issue has two parts:
(1) Ray Autoscaler: “TPU pods are nodes with multiple sub-nodes (multiple IP addresses). For example, a tpu pod v2-32 is going to have 4 IP addresses, while the current autoscaler assumes each new VM has only one IP address.” (@JiahaoYao)

(2) KubeRay: “I think the only sane way to do this would be to adapt KubeRay to work with GKE’s TPU support. This would 
only work well with statically-sized clusters (Ray autoscaler disabled).” (@DmitriGekhtman https://github.com/ray-project/ray/issues/25223#issuecomment-1364445190)

@richardsliu from GKE is interested in contributing (2). The progress is tracked by https://github.com/ray-project/kuberay/issues/908.

**Comment by infwinston** [comment]
@OrenLeung while you're waiting for the GKE support, you may try SkyPilot out. Basically one command to launch/setup/kick off a TPU Pod job.
```
sky launch examples/tpu/cifar_pod.yaml -c mypod
```
some reference: https://skypilot.readthedocs.io/en/latest/reference/tpu.html#using-tpu-pods

**Comment by brandonroyal** [comment]
Should this issue be closed because of #37934 and #38588?

**Comment by Yard1** [comment]
Think so, yeah!
